dc_mlx5/uct_flush_test.put_bcopy_flush_ep_no_comp/0 <dc_mlx5/mlx5_1:1> fails on swx-rdmz-ucx-arm-hwi

### Describe the bug
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/11933/logs/181
```
2020-12-18T10:34:25.2707303Z [ RUN      ] dc_mlx5/uct_flush_test.put_bcopy_flush_ep_no_comp/0 <dc_mlx5/mlx5_1:1>
2020-12-18T10:34:25.3440139Z [swx-rdmz-ucx-arm-hwi:58257:0:58257]  dc_mlx5_ep.c:592  Assertion `!(ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL)' failed
2020-12-18T10:34:25.3455044Z ==== backtrace (tid:  58257) ====
2020-12-18T10:34:25.3456915Z  0  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/src/ucs/.libs/libucs.so.0(ucs_fatal_error_message+0xa0) [0xffffa92db034]
2020-12-18T10:34:25.3458905Z  1  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/src/ucs/.libs/libucs.so.0(ucs_fatal_error_format+0xbc) [0xffffa92db0f4]
2020-12-18T10:34:25.3460934Z  2  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/src/uct/ib/.libs/libuct_ib.so.0(uct_dc_mlx5_ep_tag_eager_short+0) [0xffffa8eb6884]
2020-12-18T10:34:25.3462705Z  3  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x6190ec]
2020-12-18T10:34:25.3464168Z  4  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x619040]
2020-12-18T10:34:25.3465593Z  5  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x5be2c8]
2020-12-18T10:34:25.3466955Z  6  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x5a1a2c]
2020-12-18T10:34:25.3468356Z  7  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x59f178]
2020-12-18T10:34:25.3469819Z  8  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x59ee84]
2020-12-18T10:34:25.3471203Z  9  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x59ebd0]
2020-12-18T10:34:25.3472774Z 10  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x58ea78]
2020-12-18T10:34:25.3474227Z 11  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x5a289c]
2020-12-18T10:34:25.3476462Z 12  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x58e574]
2020-12-18T10:34:25.3477889Z 13  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x5aaabc]
2020-12-18T10:34:25.3479237Z 14  /lib/aarch64-linux-gnu/libc.so.6(__libc_start_main+0xe0) [0xffffa8b006e0]
2020-12-18T10:34:25.3480645Z 15  /scrap/azure/agent-08/AZP_WORKSPACE/1/s/build-test/test/gtest/gtest() [0x57bda0]
2020-12-18T10:34:25.3481479Z =================================
2020-12-18T10:34:25.3482626Z [swx-rdmz-ucx-arm-hwi:58257:0:58257] Process frozen...
2020-12-18T13:02:45.4608246Z Makefile:3354: recipe for target 'test' failed
2020-12-18T13:02:45.4609947Z make: *** [test] Terminated
2020-12-18T13:02:45.4766156Z ##[error]Bash exited with code '124'.
```

--- a/src/uct/ib/dc/dc_mlx5_ep.c
+++ b/src/uct/ib/dc/dc_mlx5_ep.c
@@ -589,9 +589,12 @@ ucs_status_t uct_dc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
 
     if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
         ucs_assert(ucs_arbiter_group_is_empty(&ep->arb_group));
-        ucs_assert(!(ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL));
         if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
             return UCS_ERR_UNSUPPORTED;
+        }
+
+        if (ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL) {
+            goto out;
         }
 
         status = uct_ib_mlx5_modify_qp_state(md, &txwq->super, IBV_QPS_ERR);
@@ -602,6 +605,7 @@ ucs_status_t uct_dc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
         ep->flags |= UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL;
     }
 
+out:
     return uct_rc_txqp_add_flush_comp(&iface->super.super, &ep->super, txqp,
                                       comp, txwq->sig_pi);
 }

--- a/src/uct/ib/rc/accel/rc_mlx5_ep.c
+++ b/src/uct/ib/rc/accel/rc_mlx5_ep.c
@@ -563,6 +563,7 @@ ucs_status_t uct_rc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
     UCT_RC_MLX5_EP_DECL(tl_ep, iface, ep);
     uct_ib_mlx5_md_t *md = ucs_derived_of(iface->super.super.super.md,
                                           uct_ib_mlx5_md_t);
+    int already_canceled = ep->super.flags & UCT_RC_EP_FLAG_FLUSH_CANCEL;
     ucs_status_t status;
     uint16_t sn;
 
@@ -585,7 +586,7 @@ ucs_status_t uct_rc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
         sn = ep->tx.wq.sig_pi;
     }
 
-    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
+    if (ucs_unlikely((flags & UCT_FLUSH_FLAG_CANCEL) && !already_canceled)) {
         status = uct_ib_mlx5_modify_qp_state(md, &ep->tx.wq.super, IBV_QPS_ERR);
         if (status != UCS_OK) {
             return status;

--- a/src/uct/ib/rc/base/rc_ep.c
+++ b/src/uct/ib/rc/base/rc_ep.c
@@ -512,7 +512,6 @@ ucs_status_t uct_rc_ep_flush(uct_rc_ep_t *ep, int16_t max_available,
 
     if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
         ucs_assert(ucs_arbiter_group_is_empty(&ep->arb_group));
-        ucs_assert(!(ep->flags & UCT_RC_EP_FLAG_FLUSH_CANCEL));
         ep->flags |= UCT_RC_EP_FLAG_FLUSH_CANCEL;
     }
 

--- a/src/uct/ib/rc/verbs/rc_verbs_ep.c
+++ b/src/uct/ib/rc/verbs/rc_verbs_ep.c
@@ -414,7 +414,8 @@ ucs_status_t uct_rc_verbs_ep_flush(uct_ep_h tl_ep, unsigned flags,
                                    uct_completion_t *comp)
 {
     uct_rc_verbs_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_verbs_iface_t);
-    uct_rc_verbs_ep_t *ep = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    uct_rc_verbs_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    int already_canceled        = ep->super.flags & UCT_RC_EP_FLAG_FLUSH_CANCEL;
     ucs_status_t status;
 
     status = uct_rc_ep_flush(&ep->super, iface->config.tx_max_wr, flags);
@@ -427,7 +428,7 @@ ucs_status_t uct_rc_verbs_ep_flush(uct_ep_h tl_ep, unsigned flags,
         uct_rc_verbs_ep_post_flush(ep, IBV_SEND_SIGNALED);
     }
 
-    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
+    if (ucs_unlikely((flags & UCT_FLUSH_FLAG_CANCEL) && !already_canceled)) {
         status = uct_ib_modify_qp(ep->qp, IBV_QPS_ERR);
         if (status != UCS_OK) {
             return status;

